Implemented Request Forwarding

pcodes opened the pull request:
## Overview
This PR adds the ability to forward incoming requests for a component to the appropriate v9_worker node that contains the matching component, and then return the response to the client. It is able to forward any type of HTTP verb (e.g. GET, POST, PUT, etc) and forwards all aspects of the request (e.g. additional variables, request body, etc).

## Limitations
- Worker node status is never updated
- Active components on worker nodes is only updated when a component does not have a worker to be forwarded to. The update function should be added to a background thread
- Worker nodes should be read in from a file or environment variables, as opposed to being hard-coded.

pcodes committed:
Added ability to check v9_worker status

pcodes committed:
Added quick debug

pcodes committed:
Fixed ignored error in main

pcodes committed:
Added request forwarding code

- NOTE: There are still some pending TODO items and clippy warnings
- Finally implemented RequestForwarder::forward_request()
- Added RwLock to the component map

pcodes committed:
Modified router to support ANY http verb

- Router no longer is hard-coded to use reqwest::get(), instead uses the
http verb that is passed in with the request
- Ran and updated code based on cargo clippy

Others commented on src/error.rs:
```diff
@@ -8,6 +8,7 @@ pub enum RouterError {
     Hyper(hyper::error::Error),
     InternalJsonHandling(serde_json::Error),
     InvalidUtf8(Utf8Error),
+    InvalidRequest(reqwest::Error),
```
Alphabetize

Others commented on src/error.rs:
```diff
@@ -51,6 +60,12 @@ impl From<serde_json::Error> for RouterError {
+        Self::InvalidRequest(e)
+    }
+}
+
```
alphabetize From implementations

pcodes replied:
By what? The return type? Or code inside the From carrots?

Others commented on src/main.rs:
```diff
@@ -40,11 +44,14 @@ fn main() {
 
-    let http_request_handler = HttpRequestHandler::new();
+    //TODO: There must be a better way to do error handling
+    let http_request_handler = HttpRequestHandler::new()?;
```
We should at least log an error here, or silently swallow errors in this phase

Others commented on src/main.rs:
```diff
@@ -18,19 +18,23 @@
 
-fn main() {
-    let log_spec = "trace, hyper=info, mio=info, tokio_reactor=info, tokio_threadpool=info";
+fn main() -> Result<(), RouterError> {
```
While we can return errors from main, perhaps we should actually handle initialization problems

Others commented on src/model.rs:
```diff
@@ -0,0 +1,122 @@
+#![allow(dead_code)]
```
Can we delete any of these structs?

Others commented on src/main.rs:
```diff
@@ -18,19 +18,23 @@
 extern crate log;
+#[macro_use]
+extern crate serde;
 
```
On line 14, delete the a `clippy::needless_pass_by_value` line -- we want that lint now!

Others commented on src/request_handler.rs:
```diff
@@ -57,14 +57,17 @@ pub fn global_request_entrypoint(
-    test_output: String,
+    //Contents of this handler is global state
+    //This might include cached status checks, a list of previously seen components, etc
+    //Modifying this state should be done with a mutex or RW lock
```
Maybe just say it's thread safe? Also, spaces at the start of comments!

pcodes replied:
You wrote that comment :P

pcodes replied:
I take it back, it was ME the whole time!

pcodes replied:
I was not under duress when I made that comment

Others commented on src/request_handler.rs:
```diff
@@ -57,14 +57,17 @@ pub fn global_request_entrypoint(
-        }
+    pub fn new() -> Result<Self, RouterError> {
+        Ok(Self {
+            request_forwarder: RequestForwarder::new()?,
```
Can we make the creation of a `RequestForwarder` infallible?

Others commented on src/router.rs:
```diff
@@ -1,4 +1,9 @@
+use hyper::{Body, Method, Response};
+use parking_lot::RwLock;
+use std::collections::HashMap;
+use std::sync::Arc;
```
Import order:

```
use std::A;
use std::B;

use hyper::A;
use hyper::B;
use parking_lot::A;
use parking_lot::B;

use crate::A;
use crate::B;
```

Others commented on src/router.rs:
```diff
@@ -27,14 +32,126 @@ impl ComponentRequest {
+pub struct WorkerNode {
+    url: String,
+    components: Vec<String>,
+    is_active: bool,
```
These fields seem pointless

pcodes replied:
Trying to remember why I thought I needed them in the first place, not sure honestly

Others commented on src/router.rs:
```diff
@@ -27,14 +32,126 @@ impl ComponentRequest {
+    pub fn new() -> Result<Self, RouterError> {
+        let mut workers = Vec::new();
+        //Why does this not need to be mutable?
+        let components_map = RwLock::new(HashMap::new());
```
It's pointless to have a mutable lock -- the lock already protects the insides!

Others commented on src/router.rs:
```diff
@@ -27,14 +32,126 @@ impl ComponentRequest {
+        //Worker Two
+        workers.push(Arc::new(WorkerNode::new(String::from(
+            "http://ec2-54-211-200-158.compute-1.amazonaws.com",
+        ))));
```
Use our nice URLs here?

pcodes replied:
Will replace with environment variables

Others commented on src/router.rs:
```diff
@@ -27,14 +32,126 @@ impl ComponentRequest {
+        };
+
+        // - scan through server list to get active components
+        res.update_workers()?;
```
If we just log on errors instead of returning the error, we can make this infallible. It's not that bad if the initial update fails anyway..

Others commented on src/router.rs:
```diff
@@ -27,14 +32,126 @@ impl ComponentRequest {
+                debug!("{}", component_name);
+                self.components_map
+                    .write()
+                    .insert(component_name.to_string(), Arc::clone(worker));
```
We probably should create a new map from scratch right? What happens if a component gets completely deactivated?

Others commented on src/router.rs:
```diff
@@ -27,14 +32,126 @@ impl ComponentRequest {
+
+        if !self.components_map.read().contains_key(&component_name) {
+            let body = Body::from("Requested component not found\n");
+            return Ok(Response::builder().status(404).body(body).unwrap());
```
Log a `warn!` here?

Others commented on src/router.rs:
```diff
@@ -27,14 +32,126 @@ impl ComponentRequest {
+        }
+
+        let url = format!(
+            "{}/sl/{}/{}?{}",
```
Did you check that the query doesn't include the question mark? Also, if there is no query there should be no question mark.

pcodes replied:
Yes, query does not contain the question mark so it's needed. And true, will remove

Others commented on Cargo.toml:
```diff
@@ -17,6 +17,7 @@ failure = { version = "0.1.5", features = ["derive"]}
+parking_lot = "0.10.0"
 serde = { version = "1.0", features = ["derive"]}
 serde_json = "1.0"
+reqwest = "0.9.22"
```
q r s

pcodes replied:
I refuse to learn the alphabet

Others commented on src/error.rs:
```diff
@@ -28,6 +29,14 @@ impl Display for RouterError {
+                    "RouterError, caused by an invalid reqwest response: {}",
+                    e
+                )?;
+            }
```
Weird formatting...

Others reviewed (changes requested):
Just a few nits

pcodes committed:
Mostly formatting fixes

- Changed line length to 105
- Previously stored components are cleared on every worker update
- Changed RequestForwarder to always succeed on creation, allowing for
  initial failure of worker node checks

pcodes committed:
Worker nodes now load from env file

Others commented on Cargo.toml:
```diff
@@ -17,6 +17,7 @@ failure = { version = "0.1.5", features = ["derive"]}
 hyper = "0.12.35"
 log = "0.4.8"
-parking_lot = "0.9.0"
+parking_lot = "0.10.0"
+reqwest = "0.9.22"
```
Nit: Have you run `cargo outdated` on this?

Others commented on src/router.rs:
```diff
@@ -27,14 +35,124 @@ impl ComponentRequest {
+        for worker in &self.workers {
+            let component_list = worker.get_component_list()?;
+            for component_name in &component_list {
+                debug!("{}", component_name);
```
Nit: Provide a little more context here?

Others commented on src/router.rs:
```diff
@@ -27,14 +35,124 @@ impl ComponentRequest {
+
+        let mut url = format!(
+            "{}/sl/{}/{}",
+            self.components_map.read().get(&component_name).unwrap().url,
```
You need to be more careful with the RwLock here. You take it to check if the entry is in the map, release it, retake it, and then assume the entry is in the map. This is a race condition

pcodes committed:
Fixed race conditions

Others commented on src/router.rs:
```diff
@@ -27,14 +35,133 @@ impl ComponentRequest {
+    pub fn update_workers(&self) -> Result<(), RouterError> {
+        let mut locked_map = self.components_map.write();
+        locked_map.clear();
+        for worker in &self.workers {
+            let component_list = worker.get_component_list()?;
+            for component_name in &component_list {
+                debug!("Adding component to map: {:?}", component_name);
+                locked_map.insert(component_name.to_string(), Arc::clone(worker));
+            }
```
Nit: After merging add a TODO to have multiple nodes per component

Others reviewed (approved):
LGTM
